library/pk.c

```diff
@@ -225,12 +225,15 @@ static inline int pk_hashlen_helper( mbedtls_md_type_t md_alg, size_t *hash_len
 {
     const mbedtls_md_info_t *md_info;
 
-    if( *hash_len != 0 )
+    if( *hash_len != 0 && md_alg == MBEDTLS_MD_NONE )
         return( 0 );
 
     if( ( md_info = mbedtls_md_info_from_type( md_alg ) ) == NULL )
         return( -1 );
 
+    if ( *hash_len != 0 && *hash_len < mbedtls_md_get_size( md_info ) )
```

Comment:
This guarantees that signature functions won't do a buffer overread when called with `hash_len = size of the hash buffer` and `hash_len` is less than the size of the hash indicated by `md_alg`. But when `hash_len` is larger, the last bytes are still ignored. So even in an LTS, it might make sense to go all the way?

Reply:
I believe the concern carried over from #4561 was that many current implementations and even the mbedtls test suites allocate `MBEDTLS_MD_MAX_SIZE` bytes for the hash buffer and for the `hash_len`. If we required the exact expected hash length then all of these use cases would fail and need to be updated. Let me know what you think.